lib/logstash/codecs/identity_map_codec.rb

```diff
@@ -42,15 +42,20 @@ class PeriodicRunner
     def initialize(listener, interval, method_symbol)
       @listener, @interval = listener, interval
       @method_symbol = method_symbol
-      @running = Concurrent::AtomicBoolean.new(false)
+      @running = Concurrent::AtomicBoolean.new(true)
     end
 
+    # The goal of periodic runner is to have a single thread to do clean up / auto flush
+    # This method is expected to access by a single thread,
+    # otherwise, multiple Thread.start could create more than one periodic runner
     def start
-      return self if running?
-      @running.make_true
+      return self unless running?
+      return self if running? && !@thread.nil?
```

Comment:
Great work tracking down this issue @kaisecheng! I might be wrong, but I think it's still can suffer race-conditions here (although it's more unlikely to happen than the previous code). There's a small chance that a thread reaches this point while the `@thread` variable is still being set by another thread. Is that correct 🤔?